openapi3filter/req_resp_decoder.go

```diff
@@ -896,127 +895,193 @@ func deepSet(m map[string]interface{}, keys []string, value interface{}) {
+	mobj := make(map[string]interface{})
 
-	for propName, propSchema := range schema.Value.Properties {
-		switch {
-		case propSchema.Value.Type.Is("array"):
-			vals := strings.Split(props[propName], urlDecoderDelimiter)
-			for _, v := range vals {
-				_, err := parsePrimitive(v, propSchema.Value.Items)
-				if err != nil {
-					return nil, handlePropParseError([]string{propName}, err)
-				}
-			}
+	for kk, value := range props {
+		keys := strings.Split(kk, urlDecoderDelimiter)
+		if strings.Contains(value, urlDecoderDelimiter) {
+			// don't support implicit array indexes anymore
+			p := pathFromKeys(keys)
+			return nil, &ParseError{path: p, Kind: KindInvalidFormat, Reason: "array items must be set with indexes"}
+		}
+		deepSet(mobj, keys, value)
+	}
+	r, err := buildResObj(mobj, nil, "", schema)
+	if err != nil {
+		return nil, err
+	}
+	result, ok := r.(map[string]interface{})
+	if !ok {
+		return nil, &ParseError{Kind: KindOther, Reason: "invalid param object", Value: result}
+	}
 
-			ivals, err := convertArrayParameterToType(vals, propSchema.Value.Items.Value.Type)
-			if err != nil {
-				return nil, handlePropParseError([]string{propName}, err)
-			}
-			obj[propName] = ivals
-		case propSchema.Value.Type.Is("object"):
-			for prop := range props {
-				if !strings.HasPrefix(prop, propName+urlDecoderDelimiter) {
-					continue
-				}
-				mapKeys := strings.Split(prop, urlDecoderDelimiter)
-				nestedSchema, err := findNestedSchema(schema, mapKeys)
-				if err != nil {
-					return nil, &ParseError{path: pathFromKeys(mapKeys), Reason: err.Error()}
-				}
-				if nestedSchema.Value.Type.Permits("array") {
-					vals := strings.Split(props[prop], urlDecoderDelimiter)
-					for _, v := range vals {
-						_, err := parsePrimitive(v, nestedSchema.Value.Items)
-						if err != nil {
-							return nil, handlePropParseError(mapKeys, err)
-						}
-					}
-					ivals, err := convertArrayParameterToType(vals, nestedSchema.Value.Items.Value.Type)
-					if err != nil {
-						return nil, handlePropParseError(mapKeys, err)
-					}
-					deepSet(obj, mapKeys, ivals)
-					continue
-				}
-				value, err := parsePrimitive(props[prop], nestedSchema)
-				if err != nil {
-					return nil, handlePropParseError(mapKeys, err)
-				}
-				deepSet(obj, mapKeys, value)
-			}
-		default:
-			value, err := parsePrimitive(props[propName], propSchema)
-			if err != nil {
-				return nil, handlePropParseError([]string{propName}, err)
-			}
-			obj[propName] = value
+	return result, nil
+}
+
+// example: map[0:map[key:true] 1:map[key:false]] -> [map[key:true] map[key:false]]
+func sliceMapToSlice(m map[string]interface{}) ([]interface{}, error) {
+	var result []interface{}
+
+	keys := make([]int, 0, len(m))
+	for k := range m {
+		key, err := strconv.Atoi(k)
+		if err != nil {
+			return nil, fmt.Errorf("array indexes must be integers: %w", err)
 		}
+		keys = append(keys, key)
 	}
 
-	return obj, nil
+	for i := 0; i <= slices.Max(keys); i++ {
```

Comment:
Ah sorry. I probably misstepped and lost the message I was leaving here.

LGTM, just: could you inline this call here? So as to drop the dependency on `x/exp`